src/VisualStudio/Core/Def/Implementation/FindUsages/VisualStudioFindSymbolMonikerUsagesService.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.Editor.SymbolMonikers;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using VS.IntelliNav.Contracts;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindUsages
+{
+    [ExportWorkspaceService(typeof(IFindSymbolMonikerUsagesService), layer: ServiceLayer.Host), Shared]
+    internal class VisualStudioFindSymbolMonikerUsagesService : AbstractFindSymbolMonikerUsagesService
+    {
+        private readonly ICodeIndexProvider? _codeIndexProvider;
+
+        [ImportingConstructor]
+        public VisualStudioFindSymbolMonikerUsagesService(
+            [Import(AllowDefault = true)] ICodeIndexProvider? codeIndexProvider)
+        {
+            _codeIndexProvider = codeIndexProvider;
+        }
+
+        public override async IAsyncEnumerable<ExternalReferenceItem> FindReferencesByMoniker(
+            DefinitionItem definition, ImmutableArray<SymbolMoniker> monikers,
+            IStreamingProgressTracker progress, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            if (_codeIndexProvider == null)
+                yield break;
+
+            var convertedMonikers = ConvertMonikers(monikers);
+            var currentPage = 0;
+            while (true)
+            {
+                var referenceItems = await FindReferencesByMonikerAsync(
+                    _codeIndexProvider, definition, convertedMonikers, progress, currentPage, cancellationToken).ConfigureAwait(false);
+
+                // If we got no items, we're done.
+                if (referenceItems.Length == 0)
+                    break;
+
+                foreach (var item in referenceItems)
+                    yield return item;
+
+                // Otherwise, we got some items.  Return them to our caller and attempt to retrieve
+                // another page.
+                currentPage++;
```

Comment:
Would it make sense to add a work item to change this behavior in a future change so that we request more pages on user demand?

Reply:
i odn't like work items being filed that i don't hav ea strong expectation will necessarily go anywhere.  either we decide that behavior is important, and we file the issue, or we can keep the status quo of being issue-less here.

Reply:
It was also discussed yesterday during a meeting that we're not sure we actually *can* know when the user has scrolled, so it'd require a few more teams to get work in before we could implement that.